### PR TITLE
Feat/finalization 0 1

### LINKED ISSUE #86: Plot direction of trips
Plot the direction of the trips

### PR DESCRIPTION
**Pull request by Speedfan94**

### CODE REVIEW

**Comment by Speedfan94** [comment]
#86

File: nextbike/visualization/geo.py [hunk 1]
```diff
@@ -1,16 +1,19 @@
 import folium
 from folium import plugins
-from nextbike import io
+from .. import io
 import datetime as dt
```
Comment (weisske) [inline nextbike/visualization/geo.py]: use importing style with from nextbike import xyz

File: nextbike/visualization/geo.py [hunk 2]
```diff
@@ -59,13 +73,11 @@ def visualize_heatmap(p_df, p_mode):
...
 
-    # ToDo: Maybe filter out 0.0 ids and duplicated places
-    # todo: variable time to plot heatmap
     m = folium.Map([49.452030, 11.076750], zoom_start=13)
 
     # mark each station as a point
     for index, row in stations.iterrows():
-        folium.CircleMarker([row["Latitude_end"], row["Longitude_end"]],
+        folium.CircleMarker([row['Latitude_end'], row['Longitude_end']],
                             radius=3,
```
Comment (weisske) [inline nextbike/visualization/geo.py]: change to "" instead ''

File: nextbike/visualization/geo.py [hunk 3]
```diff
@@ -121,30 +144,118 @@ def visualize_postalcode(p_df, p_mode):
...
         columns=["Postalcode", "Month_start"],
-        key_on="feature.properties.plz",
-        legend_name="Trips per postal code",
-        fill_color="YlGnBu",
+        key_on='feature.properties.plz',
+        legend_name='Trips per postal code',
+        fill_color='YlGnBu',
         fill_opacity=0.7,
         line_opacity=0.5,
     ).add_to(m)
```
Comment (weisske) [inline nextbike/visualization/geo.py]: Use "" instad ''

File: nextbike/visualization/geo.py [hunk 4]
```diff
@@ -121,30 +144,118 @@ def visualize_postalcode(p_df, p_mode):
...
     for index, row in df_stations.iterrows():
         folium.CircleMarker(
-            [row["Latitude_start"], row["Longitude_start"]],
+            [row['Latitude_start'], row['Longitude_start']],
             radius=3,
-            popup=[row["Place_start"], row["Latitude_start"]],
+            popup=folium.Popup(
+                "<b>Station Name:</b><br>" +
+                str(row["Place_start"]), max_width=400),
             fill_color="#3db7e4",
```
Comment (weisske) [inline nextbike/visualization/geo.py]: Use "" instead ''

File: nextbike/cli_code.py [hunk 5]
```diff
@@ -275,16 +284,17 @@ def train_direction_models(p_weather, p_mode=""):
...
+    X_scaled_train = prediction.prepare_feature.scale(p_X_train=X_train,
+                                                      p_scaler_name="Standard_Scaler_Direction" + p_weather)
     print("Do PCA...")
     # TODO: fit number of components
     X_train_transformed = prediction.prepare_feature.do_pca(p_X_scaled_train=X_scaled_train,
-                                                            p_number_components=15,
-                                                            p_filename="PCA_Direction"+p_weather)
+                                                            p_number_components=21,
+                                                            p_filename="PCA_Direction" + p_weather)
     # Train
```
Comment (weisske) [inline nextbike/cli_code.py]: try with 17 instead 21

File: nextbike/cli_code.py [hunk 6]
```diff
@@ -96,18 +98,23 @@ def training_duration_models(p_weather, p_mode=""):
...
+    components = 21
+    # if len(p_weather) > 0:
+    #     components = 21
+
     X_train_transformed = prediction.prepare_feature.do_pca(p_X_scaled_train=X_scaled_train,
-                                                            p_number_components=21,
-                                                            p_filename="PCA_Duration"+p_weather)
+                                                            p_number_components=components,
+                                                            p_filename="PCA_Duration" + p_weather)
     df_components = pd.DataFrame(X_train_transformed)
```
Comment (weisske) [inline nextbike/cli_code.py:115]: try with 17 instead 21

File: nextbike/cli.py [hunk 7]
```diff
@@ -49,10 +49,13 @@ def main(test, clean, viso, traindur, pred, traingeo, predgeo, weather):
     if weather:
         weather_data = "_weather"
     if test:
-        # df = io.read_csv("Features_Duration.csv", "output")
+        df = io.read_csv("Trips.csv", "output")
         # testing_duration_models()
```
Comment (weisske) [inline nextbike/cli.py]: Features duration is needed for testing training algorithms.  Therefore leave it in but uncommented

**Review by weisske — CHANGES_REQUESTED** [review]

**Review by weisske — APPROVED** [review]
Approve

**Pull request merged**